### ISSUE TITLE
Textblock does not respect encoding size

### ISSUE THREAD

**Issue by mrocklin** [blaze/dask#606]
### Background on `textblock`

The `textblock` function in `dask/utils.py` reads a block of bytes from a file at rough locations.  It does the following

1.  Seek to the desired start byte within the file
2.  Find the next endline
3.  Seek to the desired end byte within the file
4.  Find the next endline 
5.  Read from start endline to end endline

This lets us parallelize large files, splitting up the data cleanly between many processes.

### Problem: encoding sizes

Textblock operates on bytes, it doesn't manage encodings.  Unfortunately as is it assumes that an endline is just one byte (or two if `\r\n` is used).  This assumption fails if we use character encodings like `utf-16` or `utf-32` where all characters, including endline, fill up more than one byte.  The various `+1`s and `-1`s in `textblock` need to be modified to be aware of the encoding that will eventually be used.

This is at least partially to blame for the issues in #601

**Comment by fortiema** [comment]
Actually, reading this makes me think about my own modifications in `bag` and how data is ingested into bags' partitions. Pretty sure that the byte chunking is also not robust to multi-bytes encoding formats there, meaning the split could happen on the middle of a character for variable length encodings like UTF-8.  Need to have a second look at this!

**Comment by fortiema** [comment]
I created a branch and started hacking on this. Should be pretty easy to solve, basically add support for encoding in `textblock` so that we can pass the paramer to the `open` call.

Do we have sample data already that shows this particular problem? If not I will just create one I guess.

**Comment by mrocklin** [comment]
I was using this toned down version of the error presented in #601 for testing

```python
def test_encoding_gh601():
    with tmpfile('.csv') as fn:
        ar = pd.Series(range(0, 100))
        br = ar % 7
        cr = br * 3.3
        dr = br / 1.9836
        test_df = pd.concat([ar, br, cr, dr], axis=1)
        test_df.to_csv(fn, encoding='utf-16')

        d = dd.read_csv(fn, encoding='utf-16', dtype=str,
                        chunkbytes=1000)

        d.count().compute(get=get_sync)
```

**Comment by fortiema** [comment]
Thanks!

**Comment by cowlicks** [comment]
@fortiema how's it going?

**Comment by fortiema** [comment]
@cowlicks I've been very busy with other stuff lately, haven't had the chance to come back to this issue. Hoping I can wrap this up next weekend!

**Comment by cowlicks** [comment]
Hey @fortiema if you have not had time to pick this back up, I can work on it today.

**Comment by fortiema** [comment]
@cowlicks Please do so! I've been recently caught up in more pressing issues and didn't have time to make any significant progress on this.

**Comment by cowlicks** [comment]
Closed by #723
